README.md
# nDRandom
Tool that outputs a random coordinates generator in n-th dimensions  
GitHub: https://github.com/YujiSODE/nDRandom  
>Copyright (c) 2021 Yuji SODE \<yuji.sode@gmail.com\>  
>This software is released under the MIT License.  
>See LICENSE or http://opensource.org/licenses/mit-license.php  
______
<img width=300 src="conceptIMG_nDRandom.png" alt="conceptIMG_nDRandom">

- It can generates a random coordinates in n-th dimensions employing a single random numbers.
- Generated elements in random coordinate may show spurious correlation, but they have no correlation with each other.
- Generated elements are not cryptographically secure.

## 1. Description
`nDRandom` is a tool that outputs a random coordinates generator in n-th dimensions.  

### shell
- `tclsh nDRandom.tcl name option min1 max1 division1 ?min2 max2 division2 ?... minN maxN divisionN??;`
  - `$name`: a keyword to determine output file name and its functions
  - `$option`: a keyword to determine output type
  - `$min1, $max1 and $division1`: data range
  - `$min2, $max2 and $division2 (, ..., $minN, $maxN and $divisionN)`: optional data ranges
    - `$division`: number of divided ranges that are not less than 2

### Tcl
- `::nDRandom::setRange min max division;`: loads a variable range
- `::nDRandom::output name ?option?;`: outputs a random coordinates generater in the current directory
  - `$min and $max`: minimum and maximum values
  - `$division`: number of divided ranges that are not less than 2
  - `$name`: a keyword to determine output file and its functions
  - `$option`: an optional keyword to determine output type, and default value is `tcl`

### Available keyword for `$option`
    keyword: output type
    tcl: Tcl
    js: JavaScript

### Generated script
The generated script is output in the current directory.
- Generated script format: `${name}_info()` and `${name}_random(double)` in math functions
- `${name}_info()`: it returns preset data
- `${name}_random(double)`: it returns a random coordinates in n-th dimensions
  - `$double`: a real number in range `(0.0, 1.0)`
  - output coordinates: `{x1 dx1 x2 dx2 ... xn dxn}` where `Xi = xi+c*dxi` and `c` is a constant

## 2. [Algorithm](algorithm.md)
To obtain random coordinates from a given space in `n`-th dimensions.  
Let variable elements be _**`E`**_`= (`_`e`_<sub>`0`</sub>`,`_`e`_<sub>`1`</sub>`,...,`_`e`_*<sub>`n`</sub>*`)` where _`i`_`=0,...,`_`n`_.  
A range of a variable element `e`<sub>_`i`_</sub> is divided into _`d`_<sub>_`i`_</sub> intervals.  
Let indices correspond to a specified intervals in divided elements as follows:  
_**`IDX`**_`=(`_`IDX`_<sub>`0`</sub>`,`_`IDX`_<sub>`1`</sub>`,...,`_`IDX`_*<sub>`n`</sub>*`)`.  
  
A coordinate element value can be estimated as follows:  
_`e`_<sub>*`j`*</sub>`=`_`x`_<sub>`0`</sub>`+`_`c`_*<sub>`j`</sub>*`*`_`dx`_,  
where _`c`_*<sub>`j`</sub>* is constant,  
_`x`_<sub>`0`</sub>`=`_`e`_*<sub>`j`</sub>*`[`_`IDX`_*<sub>`j`</sub>*`]`  
and _`dx`_`=(`_`e`_*<sub>`j`</sub>*<sup>`max`</sup>`-`_`e`_*<sub>`j`</sub>*<sup>`min`</sup>`)/`_`d`_*<sub>`j`</sub>*.

## 3. Scripts
It requires Tcl 8.6+.
### main script
- [`nDRandom.tcl`](nDRandom.tcl): main script

### modules
- [`ndrandomOutput_tcl.tcl`](ndrandomOutput_tcl.tcl): module to output as Tcl file
- [`ndrandomOutput_js.tcl`](ndrandomOutput_js.tcl): module to output as JavaScript file

## 4. Library list
- Sode, Y. 2018. lSum_min.tcl: https://gist.github.com/YujiSODE/1f9a4e2729212691972b196a76ba9bd0


CHANGELOG.md
# Change Log 
## [Unreleased]

## Released: [2.1] - 2021-11-08

## Released: [2.0 beta] - 2021-11-05
## [2.0 beta] - 2021-11-05
## Fixed
- [`sample_web/sampleWeb2.js`]: fixed math functions as `ceil` to `floor`
- [`sample_web/sampleWeb.js`]: fixed math functions as `ceil` to `floor`
- [`sample02/sample02.tclory`]: fixed math functions as `ceil` to `floor`
- [`sample01/sample01.tcl`]: fixed math functions as `ceil` to `floor`
- [`ndrandomOutput_js.tcl`]: fixed math functions as `ceil` to `floor`
- [`ndrandomOutput_tcl.tcl`]: fixed math functions as `ceil` to `floor`
- [`nDRandom_IDEA.txt`]: fixed math functions as `ceil` to `floor`
- [`algorithm.md`]: fixed math functions as `ceil` to `floor`

## Released: [1.2] - 2021-08-21
## [1.2] - 2021-08-20
## Changed
- [`sample_web/index.html`]: copied from `sample_web/ndrandom_web.html`

## Added
- [`sample_web/ndrandom_web.html`] lines 73-96: a new `canvas` tag was added to draw a single triangular plane

## Changed
- [`sample_web/ndrandom_web.html`] lines 26-30:  
  `		<p>`  
  `			Tool that outputs a random coordinates generator in n-th dimensions.`  
  `			<br>This program is web sample version of <a href='https://github.com/YujiSODE/nDRandom'>nDRandom</a>.`  
  `			<br>These samples are pieces of generative art.`  
  `		</p>`
- [`sample_web/ndrandom_web.html`] line 18:  
  `		<meta name='description' content="Tool that outputs a random coordinates generator in n-th dimensions. This program is web sample version of nDRandom. These samples are pieces of generative art.">`

## Added
- [`sample_web/ndrandom_web.html`] line 14:  
  `<!-- These samples are pieces of generative art. -->`

## Changed
- [`sample_web/ndrandom_webScript.js`] line 165:  
  `	tId=setInterval(()=>{F1();F2();F3();},delay);`

## Added
- [`sample_web/ndrandom_webScript.js`] lines 98-162: a new function `F3()` was added to draw a sigle triangular plane on the `canvas` tag

## Changed
- [`sample_web/ndrandom_webScript.js`] line 94:  
  `		ctx.fill();`
- [`sample_web/ndrandom_webScript.js`] lines 88-99: removed
- [`sample_web/ndrandom_webScript.js`] line 11:  
  `	let F1=()=>{},F2=()=>{},F3=()=>{},tId=0;`

## [1.2] - 2021-08-20
## Changed
- [`sample_web/index.html`]: copied from `sample_web/ndrandom_web.html`
- [`sample_web/ndrandom_web.html`] lines 26-68: added `<div class='samples>` ... `</div>` to apply CSS

## Added
- [`sample_web/ndrandom_web_style.css`] lines 14-17:  
  `.samples{`  
  `	display:grid;`  
  `	grid-template-columns:30vw 30vw 30vw;`  
  `}`
- [`sample_web/ndrandom_web_style.css`] line 12:  
  `	overflow:auto;`

## Changed
- [`sample_web/ndrandom_web_style.css`] lines 5 and 6:  
  `	width:25vw;`  
  `	height:25vw;`

## Released: [1.1 beta] - 2021-08-18
## [1.1 beta] - 2021-08-18
## Changed
- [`sample_web/index.html`]: copied from `sample_web/ndrandom_web.html`

## Added
- [`sample_web/ndrandom_web.html`] lines 49-63: added description for random values

## Changed
- [`sample_web/ndrandom_webScript.js`] lines 80 and 81:  
  `		X3=x0+canvasW*(Rs[8]+Rs[9]*0.3)*Math.cos(Rs[14]+Rs[15]*0.3)*0.1;`  
  `		Y3=y0+canvasH*(Rs[8]+Rs[9]*0.3)*Math.sin(Rs[14]+Rs[15]*0.3)*0.1;`
- [`sample_web/ndrandom_webScript.js`] lines 76 and 77:  
  `		X2=x0+canvasW*(Rs[6]+Rs[7]*0.3)*Math.cos(Rs[12]+Rs[13]*0.3)*0.1;`  
  `		Y2=y0+canvasH*(Rs[6]+Rs[7]*0.3)*Math.sin(Rs[12]+Rs[13]*0.3)*0.1;`
- [`sample_web/ndrandom_webScript.js`] lines 72 and 73:  
  `		X1=x0+canvasW*(Rs[4]+Rs[5]*0.3)*Math.cos(Rs[10]+Rs[11]*0.3)*0.1;`  
  `		Y1=y0+canvasH*(Rs[4]+Rs[5]*0.3)*Math.sin(Rs[10]+Rs[11]*0.3)*0.1;`

## [1.1 beta] - 2021-08-16
## Changed
- [`sample_web/index.html`]: copied from `sample_web/ndrandom_web.html`
- [`sample_web/ndrandom_web.html`] line 25:  
  `		<p>Tool that outputs a random coordinates generator in n-th dimensions.<br>This program is web sample version of <a href='https://github.com/YujiSODE/nDRandom'>nDRandom</a>.</p>`
- [`sample_web/ndrandom_webScript.js`] line 42:  
  `	//------------`
- [`sample_web/ndrandom_webScript.js`] line 35:  
  `		//++++++++++++`
- [`sample_web/ndrandom_webScript.js`] lines 11 and 12:  
  `	let F1=()=>{},F2=()=>{},tId=0;`  
  `	//------------`

## Added
- [`sample_web/ndrandom_web.html`] line 21:  
  `		<script type='text/javascript' src='sampleWeb2.js'></script>`
- [`sample_web/ndrandom_web.html`] lines 43-53: a new `canvas` tag was added to draw a triangular planes
- [`sample_web/ndrandom_webScript.js`] lines 44-113: a new function `F2()` was added to draw a triangular plane on the `canvas` tag
- [`sample_web/ndrandom_sampleWeb2.tcl`]
- [`sample_web/sampleWeb2.js`]

## [1.1 beta] - 2021-08-15
## Changed
- [`sample_web/index.html`] lines 26-42: copied from `sample_web/ndrandom_web.html`
- [`sample_web/ndrandom_web.html`] lines 26-42: added `figure` tag, `figcaption` tag and figure title
- [`sample_web/ndrandom_webScript.js`] lines 40-41:  
  `	//`  
  `	tId=setInterval(()=>{F1();},delay);`
- [`sample_web/ndrandom_webScript.js`] lines 11-15:  
  `	let F1=()=>{},tId=0;`  
  `	//`  
  `	//"Random Rectangles"`  
  `	//F1 is function to draw a rectangle area on the canvas tag`  
  `	F1=()=>{`
- [`sample_web/ndrandom_web_style.css`]: a line moved from 6 to 4  

## Added
- [`sample_web/ndrandom_web_style.css`] lines 7-11:  
  ` }`  
  `figure{`  
  `	border:2px solid #f00;`  
  `	background:#eee;`  
  `	width:30vw;`

## Released: [1.01] - 2021-08-09
## [1.01] - 2021-08-06
## Added
- [`README.md`] lines 10-13:  
  ` - It can generates a random coordinates in n-th dimensions employing a single random numbers.`  
  ` - Generated elements in random coordinate may show spurious correlation, but they have no correlation with each other.`  
  ` - Generated elements are not cryptographically secure.`

## Fixed
- [`sample01/ndrandom_sample01.tcl`] lines 4-6: comments
- [`sample02/ndrandom_sample02.tcl`] lines 4-8: comments
- [`sample_web/ndrandom_sampleWeb.tcl`] lines 5-6, 9-10 and 13-15: comments

## Released: [1.0] - 2021-07-27
## [1.0] - 2021-07-27
## Fixed

sample_web/ndrandom_webScript.js
/*nDRandom
*ndrandom_webScript.js
*===================================================================
*	Copyright (c) 2021 Yuji SODE <yuji.sode@gmail.com>
*
*	This software is released under the MIT License.
*	See LICENSE or http://opensource.org/licenses/mit-license.php
*===================================================================
*/
((delay)=>{
	let F1=()=>{},F2=()=>{},F3=()=>{},tId=0;
	//------------
	//
	//"Random Rectangles"
	//F1 is function to draw a rectangle area on the canvas tag
	F1=()=>{
		let slf=window.document,C=slf.getElementById('ndrandomCvs'),ctx=C.getContext('2d'),
			Rs=Math.sampleWeb_random(Math.random()),canvasW=C.width,canvasH=C.height,
			x0=0.0,y0=0.0,w=1.0,h=1.0;
		//
		//Rs: [x1 dx1 x2 dx2 ... x7 dx7] where Xi = xi+c*dxi and c is a constant
		//Rs[0] -> x1: relative x coordinate
		//Rs[2] -> x2: relative y coordinate
		//Rs[4] -> x3: relative width
		//Rs[6] -> x4: relative height
		//Rs[8] -> x5: R value of RGB color
		//Rs[10] -> x6: G value of RGB color
		//Rs[12] -> x7: B value of RGB color
		//
		x0=canvasW*(Rs[0]+Rs[1]*0.3);
		y0=canvasH*(Rs[2]+Rs[3]*0.3);
		w=canvasW*0.1*(Rs[4]+Rs[5]*0.3);
		h=canvasH*0.1*(Rs[6]+Rs[7]*0.3);
		//
		//++++++++++++
		ctx.fillStyle=`rgb(${Math.floor(Rs[8])},${Math.floor(Rs[10])},${Math.floor(Rs[12])})`;
		//
		ctx.fillRect(x0,y0,w,h);
		//
		slf=C=ctx=Rs=canvasW=canvasH=x0=y0=w=h=null;
	};
	//------------
	//
	//"Random Triangular Planes"
	//F2 is function to draw a triangular plane on the canvas tag
	F2=()=>{
		let slf=window.document,C=slf.getElementById('ndrandomCvs2'),ctx=C.getContext('2d'),
			Rs=Math.sampleWeb2_random(Math.random()),canvasW=C.width,canvasH=C.height,
			x0=0.0,y0=0.0,X1=0.0,X2=0.0,X3=0.0,Y1=0.0,Y2=0.0,Y3=0.0;
		//
		//Rs: [x1 dx1 x2 dx2 ... x11 dx11] where Xi = xi+c*dxi and c is a constant
		//Rs[0] -> x1: relative x coordinate of an origin
		//Rs[2] -> x2: relative y coordinate of an origin
		//
		//Rs[4] -> x3: the first relative distance from the origin
		//Rs[6] -> x4: the second relative distance from the origin
		//Rs[8] -> x5: the third relative distance from the origin
		//
		//Rs[10] -> x6: the first horizontal angle
		//Rs[12] -> x7: the second horizontal angle
		//Rs[14] -> x8: the third horizontal angle
		//
		//Rs[16] -> x9: R value of RGB color
		//Rs[18] -> x10: G value of RGB color
		//Rs[20] -> x11: B value of RGB color
		//
		//P0 = (x0,y0)
		x0=canvasW*(Rs[0]+Rs[1]*0.3);
		y0=canvasH*(Rs[2]+Rs[3]*0.3);
		//
		//P1 = (X1,Y1)
		X1=x0+canvasW*(Rs[4]+Rs[5]*0.3)*Math.cos(Rs[10]+Rs[11]*0.3)*0.1;
		Y1=y0+canvasH*(Rs[4]+Rs[5]*0.3)*Math.sin(Rs[10]+Rs[11]*0.3)*0.1;
		//
		//P2 = (X2,Y2)
		X2=x0+canvasW*(Rs[6]+Rs[7]*0.3)*Math.cos(Rs[12]+Rs[13]*0.3)*0.1;
		Y2=y0+canvasH*(Rs[6]+Rs[7]*0.3)*Math.sin(Rs[12]+Rs[13]*0.3)*0.1;
		//
		//P3 = (X3,Y3)
		X3=x0+canvasW*(Rs[8]+Rs[9]*0.3)*Math.cos(Rs[14]+Rs[15]*0.3)*0.1;
		Y3=y0+canvasH*(Rs[8]+Rs[9]*0.3)*Math.sin(Rs[14]+Rs[15]*0.3)*0.1;
		//
		//++++++++++++
		ctx.fillStyle=`rgb(${Math.floor(Rs[16])},${Math.floor(Rs[18])},${Math.floor(Rs[20])})`;
		//
		//triangular plane: (P1,P2,P3)
		//P1 -> P2, P2 -> P3, P3 -> P1
		//
		ctx.beginPath();
		ctx.moveTo(X1,Y1);
		ctx.lineTo(X2,Y2);
		ctx.lineTo(X3,Y3);
		ctx.closePath();
		ctx.fill();
		//
		slf=C=ctx=Rs=canvasW=canvasH=x0=y0=X1=X2=X3=Y1=Y2=Y3=null;
	};
	//------------
	//
	//"Random Single Triangular Plane"
	//F3 is function to draw a sigle triangular plane on the canvas tag
	F3=()=>{
		let slf=window.document,C=slf.getElementById('ndrandomCvs3'),ctx=C.getContext('2d'),
			Rs=Math.sampleWeb2_random(Math.random()),canvasW=C.width,canvasH=C.height,
			x0=canvasW*0.5,y0=canvasH*0.5,X1=0.0,X2=0.0,X3=0.0,Y1=0.0,Y2=0.0,Y3=0.0;
		//
		//Rs: [x1 dx1 x2 dx2 ... x11 dx11] where Xi = xi+c*dxi and c is a constant
		//Rs[0] -> x1: relative x coordinate of an origin (= width/2)
		//Rs[2] -> x2: relative y coordinate of an origin (= height/2)
		//
		//Rs[4] -> x3: the first relative distance from the origin
		//Rs[6] -> x4: the second relative distance from the origin
		//Rs[8] -> x5: the third relative distance from the origin
		//
		//Rs[10] -> x6: the first horizontal angle
		//Rs[12] -> x7: the second horizontal angle
		//Rs[14] -> x8: the third horizontal angle
		//
		//Rs[16] -> x9: R value of RGB color
		//Rs[18] -> x10: G value of RGB color
		//Rs[20] -> x11: B value of RGB color
		//
		//P1 = (X1,Y1)
		X1=x0+canvasW*(Rs[4]+Rs[5]*0.3)*Math.cos(Rs[10]+Rs[11]*0.3)*0.1;
		Y1=y0+canvasH*(Rs[4]+Rs[5]*0.3)*Math.sin(Rs[10]+Rs[11]*0.3)*0.1;
		//
		//P2 = (X2,Y2)
		X2=x0+canvasW*(Rs[6]+Rs[7]*0.3)*Math.cos(Rs[12]+Rs[13]*0.3)*0.1;
		Y2=y0+canvasH*(Rs[6]+Rs[7]*0.3)*Math.sin(Rs[12]+Rs[13]*0.3)*0.1;
		//
		//P3 = (X3,Y3)
		X3=x0+canvasW*(Rs[8]+Rs[9]*0.3)*Math.cos(Rs[14]+Rs[15]*0.3)*0.1;
		Y3=y0+canvasH*(Rs[8]+Rs[9]*0.3)*Math.sin(Rs[14]+Rs[15]*0.3)*0.1;
		//
		//++++++++++++
		ctx.clearRect(0,0,canvasW,canvasH);
		ctx.fillStyle=`rgb(${Math.floor(Rs[16])},${Math.floor(Rs[18])},${Math.floor(Rs[20])})`;
		//
		//triangular plane: (P1,P2,P3)
		//P1 -> P2, P2 -> P3, P3 -> P1
		//
		//### to show projections ###
		ctx.strokeStyle=`rgb(${Math.floor(Rs[16])},${Math.floor(Rs[18])},${Math.floor(Rs[20])})`;
		ctx.beginPath();
		ctx.moveTo(x0,y0);
		ctx.lineTo(X1,Y1);
		ctx.moveTo(x0,y0);
		ctx.lineTo(X2,Y2);
		ctx.moveTo(x0,y0);
		ctx.lineTo(X3,Y3);
		ctx.stroke();
		//##########################
		//
		ctx.beginPath();
		ctx.moveTo(X1,Y1);
		ctx.lineTo(X2,Y2);
		ctx.lineTo(X3,Y3);
		ctx.closePath();
		ctx.fill();
		//
		slf=C=ctx=Rs=canvasW=canvasH=x0=y0=X1=X2=X3=Y1=Y2=Y3=null;
	};
	//------------
	//
	tId=setInterval(()=>{F1();F2();F3();},delay);
})(400);
